src/main/java/org/mozilla/jss/pkcs11/PK11Signature.java

```diff
@@ -295,21 +325,30 @@ public int engineSign(byte[] outbuf, int offset, int len)
     public boolean engineVerify(byte[] sigBytes)
         throws SignatureException, TokenException
     {
-        assert(sigBytes!=null);
-		if(state != VERIFY) {
-			throw new SignatureException(
-						"Signature is not initialized properly");
-		}
-		if(!raw && sigContext==null) {
-			throw new SignatureException("Signature has no context");
-		}
-        if(raw && rawInput==null) {
-            throw new SignatureException("Signature has no input");
+        if (sigBytes == null) {
+            throw new SignatureException("No signature bytes provided");
+        }
+        if(state != VERIFY) {
+            throw new SignatureException("Signature is not initialized properly");
+            }
+        if(!raw && sigContext == null) {
+            throw new SignatureException(NO_CONTEXT);
+            }
+        if(raw && rawInput == null) {
+            throw new SignatureException(NO_INPUT_STREAM);
+        }
+        if (token == null) {
+            throw new TokenException(NO_TOKEN);
+        }
+        if (tokenProxy == null) {
+            throw new TokenException(NO_TOKEN_PROXY);
+        }
+        if (algorithm == null) {
+            throw new SignatureException(NO_ALGORTIHM);
+        }
```

Comment:
Similarly, no need to check these fields again here.

Reply:
Ack.